src/main/java/com/echobox/api/linkedin/connection/versioned/VersionedVideoConnection.java

```diff
@@ -34,13 +57,96 @@ public class VersionedVideoConnection extends VersionedConnection {
   
   /**
    * Instantiates a new connection base.
    *
    * @param linkedinClient the LinkedIn client
    */
-  protected VersionedVideoConnection(VersionedLinkedInClient linkedinClient) {
+  public VersionedVideoConnection(VersionedLinkedInClient linkedinClient) {
     super(linkedinClient);
   }
+  
+  public URN uploadVideo(InitializeUploadRequest initializeUploadRequest, String filePath)
+      throws IOException {
+  
+    Path videoFilePath = Paths.get(filePath);
+    long fileSizeBytes = Files.size(videoFilePath);
+    ValidationUtils.validateVideoFile(fileSizeBytes);
+  
+    initializeUploadRequest.getInitializeUploadRequest().setFileSizeBytes(fileSizeBytes);
+    InitializeUploadResponse initializeUploadResponse = initializeUpload(initializeUploadRequest);
+    InitializeUploadResponse.Value value = initializeUploadResponse.getValue();
+    
+    List<String> uploadedPartIds = uploadVideoFile(filePath, value.getUploadInstructions());
+    
+    FinalizeUploadRequest finalizeUploadRequest =
+        new FinalizeUploadRequest(value.getVideo(), value.getUploadToken(), uploadedPartIds);
+    finalizeUpload(finalizeUploadRequest);
+    
+    return value.getVideo();
+  }
+  
+  public InitializeUploadResponse initializeUpload(
+      InitializeUploadRequest initializeUploadRequest) {
+    return linkedinClient.publish(VIDEOS, InitializeUploadResponse.class, initializeUploadRequest,
+        Parameter.with(ACTION_KEY, INITIALIZE_UPLOAD));
+  }
+  
+  public List<String> uploadVideoFile(String filePath,
+      List<InitializeUploadResponse.UploadInstruction> uploadInstructions) throws IOException {
+    
+    File file = new File(filePath);
+    byte[] fileBytes = convertToBytes(file);
+    
+    List<String> uploadPartIds = new ArrayList<>();
+    for (InitializeUploadResponse.UploadInstruction instruction : uploadInstructions) {
+      String etag = uploadVideoFileChunk(filePath, fileBytes, instruction);
+      uploadPartIds.add(etag);
+    }
+    
+    return uploadPartIds;
+  }
+  
+  public String uploadVideoFileChunk(String filePath, byte[] fileBytes,
+      InitializeUploadResponse.UploadInstruction instruction) throws IOException {
+    WebRequestor webRequestor = linkedinClient.getWebRequestor();
+    
+    Map<String, String> requestHeaders = new HashMap<>();
+    requestHeaders.put(DefaultVersionedLinkedInClient.HEADER_NAME_VERSION,
+        DefaultVersionedLinkedInClient.DEFAULT_VERSIONED_MONTH);
+    
+    byte[] chunkBytes = Arrays.copyOfRange(fileBytes,
+        Math.toIntExact(instruction.getFirstByte()),
+        Math.toIntExact(instruction.getLastByte() + 1));
+    BinaryAttachment attachment = BinaryAttachment.with(filePath, chunkBytes,
+        ContentType.APPLICATION_OCTET_STREAM.toString());
+    
+    URL url = new URL(instruction.getUploadUrl());
+    WebRequestor.Response response =
+        webRequestor.executePut(url.toString(), null, null, requestHeaders, attachment);
```

Comment:
May I ask why we need to create a URL for instruction.getUploadUrl(), and then convert it back to String by toString()? instead of just call executePut(instruction.getUploadUrl(), ... ?

Reply:
I instantiated a `URL` object because it has some validations to ensure that the url that we receive back is correctly formed. We should never run into that problem, but you never know.

Reply:
Good point. I suggest to wrap this in a try-catch block, and then throw a more precise error. Say,` "Invalid uploadUrl returned from LinkedIn" `as a `LinkedInResponseException`?

Reply:
Sounds like a good suggestion. Will implement that